### PR TITLE
Components: add GPIO, LED, AlarmMux, UartMux

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request adds several components. I am using these for the Nano 33 board, but no need to wait to merge these in case other boards want to use them.

As an aside, my goal is to componentize everything that the Nano 33 uses.

### Testing Strategy

This pull request was tested by getting the Nano 33 board up and running.

### TODO or Help Wanted

I mean if someone wants to figure out how to have variable sized arrays in components that would be useful here, but we shouldn't block on that at this point.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
I think we should merge #1517 into this PR. I don't think it makes sense to merge the hardcoded length versions and them change them later since we have an example of how to do variable length components.

**Comment by hudson-ayers** [comment]
I'll do that.

**Comment by hudson-ayers** [comment]
Okay, I replaced the original GPIO/LED commit with the variable sized versions instead. I also added a generic ButtonComponent. In the process I took the variable size macro approach a bit further and I think it is much cleaner to use now in every case (though the helper macros themselves are a bit difficult to parse). I also modified main.rs for Imix and Hail to use these components, and tested on Imix that everything still works as expected with the console/buttons/leds/gpio. 

This PR is ready for a fresh review.

File: kernel/src/common/utils.rs [hunk 1]
```diff
@@ -90,3 +90,19 @@ macro_rules! create_capability {
...
+/// Useful for constructing variable sized arrays in other macros.
+/// Taken from the Little Book of Rust Macros
+///
+/// ```ignore
+/// use kernel:count_exprs;
+///
+/// let count: usize = count_exprs!(1+2, 3+4);
+/// ```
+#[macro_export]
+macro_rules! count_exprs {
```
Comment (bradjc) [inline kernel/src/common/utils.rs]: I would prefer this to be `count_expressions` but I can understand if it is better to match the Rust shorthand.

**Comment by bradjc** [comment]
Aha! #1151 is a dream a year and a half in the making, but I think we are finally there! I don't think I can approve my own PR but this looks great to me.

File: kernel/src/common/utils.rs (hunk 1 above)
Comment (hudson-ayers) [reply kernel/src/common/utils.rs]: I don't particularly care either way, count_expressions it is

**Comment by hudson-ayers** [comment]
(force pushed just to change `count_exprs` --> `count_expressions` as though it had originally been `count_expressions`)

And if you approve this PR I will use my approval as a proxy!

**Comment by phil-levis** [comment]
+bors

**Pull request merged**